### ISSUE TITLE
SSL Handshake failure when trying to connect to Mutual TLS secured server

### ISSUE THREAD

**Issue by victorenr-bit** [haproxy/haproxy#861]
## Detailed description of the problem

HAProxy is not able to negotiate a secure connection to a Mutual TLS secured server. What I am trying to achieve is emulate the grpc_ssl_certificate and grpc_ssl_key directives from nginx in haproxy, so basically I am trying to make the client part of HAProxy authenticate against my backend, allowing other internal services to communicate with HAProxy directlly without having to provide a certificate, so just to clarify it again what's it's failing it's the connection between HAProxy and the server.

Health checks also return the same response, this is a copy paste of the log lines:
reason: Layer6 invalid response, info: “SSL handshake failure”, check duration: 1ms, status: 0/2 DOWN.

When I try to send a request to the server I get:
10.119.17.138:52965 [16/Sep/2020:10:01:39.640] fqdn/1: SSL handshake failure

Also this is the config I am using for this particular backend:

```
global
  maxconn 50000
  log stdout format raw local0 debug

defaults
  log global
  maxconn 3000
  mode http
  timeout client          30s
  timeout server          30s
  timeout connect         10s
  option httplog
  option logasap

frontend frontend-1
  bind *:port ssl crt /letsencrypt.pem alpn h2
  mode http
  default_backend backend_internal

backend backend_internal
  balance roundrobin
  retries 3
  option redispatch
  option forwardfor
  option log-health-checks
  option http-use-htx
  server s1 fqdn:port ssl crt /client.pem verify required ca-file /ca.pem alpn h2 check check-ssl
  server s2 fqdn:port ssl crt /client.pem verify required ca-file /ca.pem alpn h2 check check-ssl
```
client.pem contains both the certificate and the key, and ca.pem is the public certificate of the CA used to generate the other certificate (Both are working with the previous nginx directives). Also just to clarify:

$ openssl verify -CAfile ca.pem client.pem 
client.pem: OK

And I can connect to the server using:
$ openssl s_client -connect fqdn:port -cert client.pem -key client.pem -CAfile ca.pem

```
...
Acceptable client certificate CA names
CN = Test CA
Requested Signature Algorithms: RSA-PSS+SHA256:ECDSA+SHA256:Ed25519:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA384:ECDSA+SHA512:RSA+SHA1:ECDSA+SHA1
Shared Requested Signature Algorithms: RSA-PSS+SHA256:ECDSA+SHA256:Ed25519:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA384:ECDSA+SHA512
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 1721 bytes and written 3073 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
...
```
Note though that the certificate CN doesn't match the server fqdn, but this is not an issue for Nginx at all.

## Expected behavior

HAProxy should be able to connect to the server.

## Steps to reproduce the behavior
Start the HAProxy service, and wait for the health checks to mark the servers as down, or remove the health checks and send traffic to be forwarded to the server.

1. 
2. 
3. 

## What is your configuration?

See the snippet I put on the first section.

```
(paste your output here)
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.2.3-0e58a34 2020/09/08 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.3.html
Running on: Linux 4.15.0-1058-kvm #59-Ubuntu SMP Thu Mar 26 01:26:50 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.0
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app
```

## If HAProxy crashed: Last outputs and backtraces

```
(paste your output here)
```

**Comment by EmericBr** [comment]
I understand you notice handshake failure on health check but what about normal traffic? Did you try connections removing health check (removing/comment check and check-ssl on server lines)

**Comment by capflam** [comment]
It does not seems to be specific to the healthchecks because the same happens sending a request manually. But here the SSL handshake may fail because of haproxy, if it reject the server certificate. Maybe because of the CN. I don't know. Could you add `no-verifyhost` option on the server line. If it still fail, you may try to set `verify` option to `none` (for test purpose).

**Comment by victorenr-bit** [comment]
I will try today with no-verifyhost and let you know if that fixes the issue, also just to clarify it, I get the same error message when doing a manual request.

**Comment by victorenr-bit** [comment]
> I understand you notice handshake failure on health check but what about normal traffic? Did you try connections removing health check (removing/comment check and check-ssl on server lines)

Yeah I tried removing the health check but the result is the same.

> It does not seems to be specific to the healthchecks because the same happens sending a request manually. But here the SSL handshake may fail because of haproxy, if it reject the server certificate. Maybe because of the CN. I don't know. Could you add `no-verifyhost` option on the server line. If it still fail, you may try to set `verify` option to `none` (for test purpose).

no-verifyhost is not a valid option here, I get unknown keyword if I try to set that directive. I tried verify none though, and that gives me the same result for both the health check and manual requests "SSL handshake failure".

**Comment by EmericBr** [comment]
I notice that the ca.pem used signed the client.pem, It means you should add it to the trusted ca on your servers, to validate client.pem. On Haproxy's side, the ca-file on server line is there to validate the server's certificate, so in your case the server certificate should be signed using the same ca. Could you confirm?

verify none disables verify on haproxy side, so if you continue to reach this issue setting verify to none, it means, if the issue is due to certificate authentication: this your server which refuses the the client.pem submitted by haproxy. To check why you should check on the server logs (or perhaps use openssl s_server to debug why).

In an other hand, this issue may not be related to certificate authentication but something else, did you try to disable the cert authentication on the servers, for test?

**Comment by victorenr-bit** [comment]
> I notice that the ca.pem used signed the client.pem, It means you should add it to the trusted ca on your servers, to validate client.pem. On Haproxy's side, the ca-file on server line is there to validate the server's certificate, so in your case the server certificate should be signed using the same ca. Could you confirm?

The server does have the same CA installed, and Mutual TLS is working with nginx like I stated at the beginning, so as far as I can tell there is no issue with the certificates, in fact if you check the provided information I am able to negotiate a connection with it using openssl s_client. I am using the exact same files nginx is using (but I combined the key with the cert since HAProxy requires it).
 
> verify none disables verify on haproxy side, so if you continue to reach this issue setting verify to none, it means, if the issue is due to certificate authentication: this your server which refuses the the client.pem submitted by haproxy. To check why you should check on the server logs (or perhaps use openssl s_server to debug why).

openssl s_client works against the actual server (both key and cert are on the client.pem file):
openssl s_client -connect fqdn:port -cert client.pem -key client.pem -CAfile ca.pem

> In an other hand, this issue may not be related to certificate authentication but something else, did you try to disable the cert authentication on the servers, for test?

The same service is also listening on a non secured port and HAProxy is able to talk/connect to it, the issue is just related with the mtls setup as far as I can tell you.

PS: I fixed the openssl s_client line on my initial report I forgot to rename the file.

**Comment by EmericBr** [comment]
> The server does have the same CA installed, and Mutual TLS is working with nginx like I stated at the beginning, so as far as I can tell there is no issue with the certificates, in fact if you check the provided information I am able to negotiate a connection with it using openssl s_client. I am using the exact same files nginx is using (but I combined the key with the cert since HAProxy requires it).


> > verify none disables verify on haproxy side, so if you continue to reach this issue setting verify to none, it means, if the issue is due to certificate authentication: this your server which refuses the the client.pem submitted by haproxy. To check why you should check on the server logs (or perhaps use openssl s_server to debug why).
> 
> openssl s_client works against the actual server (both key and cert are on the client.pem file):
> openssl s_client -connect fqdn:port -cert client.pem -key client.pem -CAfile ca.pem
> 
> > In an other hand, this issue may not be related to certificate authentication but something else, did you try to disable the cert authentication on the servers, for test?
> 
> The same service is also listening on a non secured port and HAProxy is able to talk/connect to it, the issue is just related with the mtls setup as far as I can tell you.
> 
> PS: I fixed the openssl s_client line on my initial report I forgot to rename the file.

**Comment by EmericBr** [comment]
Mistakenly closed

**Comment by EmericBr** [comment]
> The same service is also listening on a non secured port and HAProxy is able to talk/connect to it, the issue is just related with the mtls setup as far as I can tell you.
> 
But did you try disabling mtls setup but still using tls?

**Comment by victorenr-bit** [comment]
Not yet, unfortunately the server is in production, so I cannot test this kind of stuff on it right now. Anyway if I cannot use mtls there is no point to continue with HAProxy at all. The server has hardcoded the use of mtls so it is a requirement:

```
...
// Create the gRPC server
	gs.server = grpc.NewServer([]grpc.ServerOption{
		// Sets the message size in bytes the server can receive
		grpc.MaxRecvMsgSize(int(gs.conf.Server.MaxMessageSize)),
		// Set TLS options
		grpc.Creds(credentials.NewTLS(&tls.Config{
			ClientAuth:   tls.RequireAndVerifyClientCert,
			Certificates: []tls.Certificate{certificate},
			ClientCAs:    certPool,
		})),
...
```

**Comment by capflam** [comment]
Because `verify none` on the server line does not work, it means it is the connection is rejected by your server. Do you have any log message on your server side that may explain the failure ?

**Comment by victorenr-bit** [comment]
I am going to try to enable debug mode and try to get some log messages from the server, since at the current level I don't get any details about why the connection is rejected. I will get back to you later.

**Comment by EmericBr** [comment]
I've checked the reg test about ssl authentication, and did additional tests, I didn't notice any issue. I'm now quite convinced your issue is not related to cert authentication.

**Comment by victorenr-bit** [comment]
> I've checked the reg test about ssl authentication, and did additional tests, I didn't notice any issue. I'm now quite convinced your issue is not related to cert authentication.

Do you have any idea about what the issue could be then?

I have enabled debug on the server but it is not showing any extra information with regards to HAProxy, note though, that if I do a faulty connection with openssl s_client without providing the certificates it shows a message stating that the ssl negotiation failed, I do not get this message when I route a request through HAProxy, but I get the error in the HAProxy log. 

I have running both nginx and haproxy now running on different ports and routing traffic to the same service, haproxy still doesn't work but at least I am in a better position to test now. I can run tcpdump on the service side to see if there is any difference in the SSL negotiation, though I am not an expert examining network traffic and probably I would need some guidance.

**Comment by victorenr-bit** [comment]
Ok, after doing some more tests I finally got it to work. I don't understand what has changed exactly, but now even the health checks are working. I have customized the docker image to include our custom certs instead of bind mounting them inside the container and that seems to have fixed it for now, it's either that or some change on the service side that I am not aware of yet, anyway it seems to be working now.

I am marking the issue as closed, again thanks for your help!.

Thanks for your help guys.
